File: scripts/package/disease_phenotype_association_util.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
import xml.etree.ElementTree as ET

import duckdb
from rdflib import Graph, Literal, Namespace, URIRef, BNode
from rdflib.namespace import DCTERMS, RDF, RDFS, FOAF

from utils.log_util import get_logger
from package.rdf_build_support import (
    open_text_writer,
)

logger = get_logger()

HPOA_SOURCE_URI = (
    "http://compbio.charite.de/jenkins/job/hpo.annotations.current/"
    "lastSuccessfulBuild/artifact/current/phenotype.hpoa"
)

ORDO_FREQUENCY_TO_HPO = {
    "Obligate (100%)": "0040280",
    "Very frequent (99-80%)": "0040281",
    "Frequent (79-30%)": "0040282",
    "Occasional (29-5%)": "0040283",
    "Very rare (<4-1%)": "0040284",
    "Excluded (0%)": "0040285",
}

OA = Namespace('http://www.w3.org/ns/oa#')
OBO = Namespace("http://purl.obolibrary.org/obo/")
HOOM = Namespace("http://www.semanticweb.org/ontology/HOOM#")
ORDO = Namespace("http://www.orpha.net/ORDO/")


@dataclass
class AnnotationSource:
    creator: str
    page: str


@dataclass
class OrdoPhenotypeAnnotation:
    ordo_id: str
    hpo_id: str
    frequency_term_id: str | None


def load_ordo_frequency_annotations(orphanet_product4_path: str | Path) -> dict[str, str]:
    logger.info("loading Orphanet frequency annotations: path=%s", orphanet_product4_path)
    frequencies: dict[str, str] = {}
    tree = ET.parse(orphanet_product4_path)
    root = tree.getroot()

    for disorder_list in root.iter("HPODisorderSetStatusList"):
        for disorder in disorder_list.iter("Disorder"):
            orpha_code = disorder.find(".//OrphaCode")
            if orpha_code is None or orpha_code.text is None:
                continue

            for association in disorder.findall(".//HPODisorderAssociation"):
                hpo_id_element = association.find(".//HPOId")
                if hpo_id_element is None or hpo_id_element.text is None:
                    continue

                hpo_frequency_element = association.find(".//HPOFrequency")
                frequency_label = extract_frequency_label(hpo_frequency_element)
                if frequency_label is None:
                    continue

                key = f"{orpha_code.text.strip()}\t{normalize_hpo_id(hpo_id_element.text)}"
                frequencies.setdefault(key, frequency_label)

    logger.info("loaded Orphanet frequency annotations: path=%s annotations=%s", orphanet_product4_path, len(frequencies))
    return frequencies


def load_manual_phenotype_associations(
    phenotype_hpoa_path: str | Path,
    disease_prefix: str,
) -> dict[str, str]:
    logger.info(
        "loading manual phenotype associations: path=%s disease_prefix=%s",
        phenotype_hpoa_path,
        disease_prefix,
    )
    manual_associations: dict[str, str] = {}
    prefix = f"{disease_prefix}:"

    con = duckdb.connect()
    query_statement = f"""
        select
            replace(database_id, '{prefix}', ''),
            replace(hpo_id, 'HP:', '')
        from
            read_csv('{phenotype_hpoa_path}', delim='\t')
        where
            prefix(database_id, '{prefix}')
        """
    res = con.execute(query_statement)

    while True:
        row = res.fetchone()
        if row is None:
            break
        disease_id = row[0]
        hpo_id = row[1]
        key = f"{disease_id}\t{hpo_id}"
        manual_associations.setdefault(key, "Manual")

    logger.info(
        "loaded manual phenotype associations: path=%s disease_prefix=%s associations=%s",
        phenotype_hpoa_path,
        disease_prefix,
        len(manual_associations),
    )
    return manual_associations

def extract_frequency_label(hpo_frequency_element: ET.Element | None) -> str | None:
    if hpo_frequency_element is None:
        return None

    name = hpo_frequency_element.find("Name")
    if name is None or name.text is None:
        return None
    return name.text.strip()


def normalize_hpo_id(value: str) -> str:
    return value.strip().replace("HP:", "")


def create_annotation_source(creator: str, page: str) -> AnnotationSource:
    logger.info("creating annotation source: creator=%s page=%s", creator, page)
    return AnnotationSource(creator=creator, page=page)


def write_ordo_phenotype_association_ttl(
    output_path: str | Path,
    manual_associations: dict[str, str],
    frequency_by_association: dict[str, str],
    source: AnnotationSource,
) -> None:
    logger.info(
        "writing Orphanet phenotype association TTL: output=%s manual_associations=%s frequencies=%s",
        output_path,
        len(manual_associations),
        len(frequency_by_association),
    )
    graph = Graph()
    graph.bind("dcterms", DCTERMS)
    graph.bind("foaf", FOAF)
    graph.bind("hoom", HOOM)
    graph.bind("oa", OA)
    graph.bind("obo", OBO)
    graph.bind("ordo", ORDO)
    graph.bind("rdf", RDF)
    graph.bind("rdfs", RDFS)

    blank_node_counter = 0
    for annotation in build_ordo_annotations(manual_associations, frequency_by_association):
        blank_node_counter += 1
        source_node = BNode(f"b{blank_node_counter}")

        disease = URIRef(
            "https://pubcasefinder.dbcls.jp/phenotype_context/"
            f"disease:ORDO:{annotation.ordo_id}/phenotype:HP:{annotation.hpo_id}"
        )

        graph.add((disease, RDF.type, OA.Annotation))
        graph.add((disease, OA.hasTarget, ORDO[f"Orphanet_{annotation.ordo_id}"]))
        graph.add((disease, OA.hasBody, OBO[f"HP_{annotation.hpo_id}"]))
        if annotation.frequency_term_id is not None:
            graph.add((disease, HOOM.with_frequency, OBO[f"HP_{annotation.frequency_term_id}"]))
        graph.add((disease, DCTERMS.source, source_node))
        graph.add((disease, OBO["ECO_9000001"], OBO["ECO_0000218"]))

        graph.add((source_node, DCTERMS.creator, Literal(source.creator)))
        graph.add((source_node, FOAF.page, URIRef(source.page)))

    for label, hpo_id in ORDO_FREQUENCY_TO_HPO.items():
        graph.add((OBO[f"HP_{hpo_id}"], RDFS.label, Literal(label, lang="en")))

    with open_text_writer(output_path) as writer:
        writer.write(graph.serialize(format='turtle'))
    logger.info("finished writing Orphanet phenotype association TTL: output=%s triples=%s", output_path, len(graph))


def write_manual_phenotype_association_ttl(
    output_path: str | Path,
    disease_namespace_in_path: str,
    disease_resource_prefix: str,
    disease_resource_prefix_uri: str,
    manual_associations: dict[str, str],
    source: AnnotationSource,
) -> None:

    logger.info(
        "writing manual phenotype association TTL: output=%s disease_namespace=%s associations=%s",
        output_path,
        disease_namespace_in_path,
        len(manual_associations),
    )
    disease_namespace = Namespace(disease_resource_prefix_uri)

    graph = Graph()
    graph.bind("dcterms", DCTERMS)
    graph.bind("foaf", FOAF)
    graph.bind(disease_resource_prefix, disease_namespace)
    graph.bind("oa", OA)
    graph.bind("obo", OBO)
    graph.bind("rdf", RDF)
    graph.bind("rdfs", RDFS)

    blank_node_counter = 0
    for key in manual_associations:
        disease_id, hpo_id = key.split("\t")
        blank_node_counter += 1
        source_node = BNode(f"b{blank_node_counter}")
        source_creator = source.creator
        source_page = source.page

        disease = URIRef(f'https://pubcasefinder.dbcls.jp/phenotype_context/disease:{disease_namespace_in_path}:{disease_id}/phenotype:HP:{hpo_id}')

        graph.add((disease, RDF.type, OA.Annotation))
        graph.add((disease, OA.hasTarget, disease_namespace[disease_id]))
        graph.add((disease, OA.hasBody, OBO[f'HP_{hpo_id}']))
        graph.add((disease, DCTERMS.source, source_node))
        graph.add((disease, OBO['ECO_9000001'], OBO['ECO_0000218']))

        graph.add((source_node, DCTERMS.creator, Literal(source_creator)))
        graph.add((source_node, FOAF.page, URIRef(source_page)))
    with open_text_writer(output_path) as writer:
        writer.write(graph.serialize(format='turtle'))
    logger.info("finished writing manual phenotype association TTL: output=%s triples=%s", output_path, len(graph))

def build_ordo_annotations(
    manual_associations: dict[str, str],
    frequency_by_association: dict[str, str],
) -> list[OrdoPhenotypeAnnotation]:
    logger.info(
        "building Orphanet phenotype annotations: manual_associations=%s frequencies=%s",
        len(manual_associations),
        len(frequency_by_association),
    )
    annotations: list[OrdoPhenotypeAnnotation] = []
    for key in manual_associations:
        ordo_id, hpo_id = key.split("\t")
        frequency_label = frequency_by_association.get(key)
        annotations.append(
            OrdoPhenotypeAnnotation(
                ordo_id=ordo_id,
                hpo_id=hpo_id,
                frequency_term_id=ORDO_FREQUENCY_TO_HPO.get(frequency_label),
            )
        )
    logger.info("built Orphanet phenotype annotations: annotations=%s", len(annotations))
    return annotations


def write_frequency_labels(writer) -> None:
    logger.info("writing Orphanet frequency labels: labels=%s", len(ORDO_FREQUENCY_TO_HPO))
    for label, hpo_id in ORDO_FREQUENCY_TO_HPO.items():
        writer.write(f"obo:HP_{hpo_id}\n")
        writer.write(f'    rdfs:label "{label}"@en .\n')
    logger.info("finished writing Orphanet frequency labels")

File: scripts/java/DiseasePhenotypeAssociationUtil.java
package pubcasefinder_260415;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Properties;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

// 질환-표현형 연관 RDF 생성에 필요한 공통 로직을 모은 유틸리티
public class DiseasePhenotypeAssociationUtil {
	private static final Properties CONFIG = RdfBuildSupport.loadConfig();
	public static final String ORPHANET_PRODUCT4_PATH = RdfBuildSupport.resolveConfiguredFile(CONFIG, "orphanet.product4.path", RdfBuildSupport.resolveResourceRoot(CONFIG, "orphanet.dir"), "en_product4.xml");
	public static final String HPO_PHENOTYPE_PATH = RdfBuildSupport.resolveConfiguredFile(CONFIG, "hpo.phenotype.path", RdfBuildSupport.resolveResourceRoot(CONFIG, "hpo.dir"), "phenotype.hpoa");
	public static final String RDF_DIR = RdfBuildSupport.resolveConfiguredOutputDir(CONFIG);
	public static final String HPOA_SOURCE_URI = "http://compbio.charite.de/jenkins/job/hpo.annotations.current/lastSuccessfulBuild/artifact/current/phenotype.hpoa";

	private static final LinkedHashMap<String, String> ORDO_FREQUENCY_TO_HPO = createOrdoFrequencyMap();

	public static class AnnotationSource {
		String creator;
		String page;
	}

	private static class OrdoPhenotypeAnnotation {
		String ordoId;
		String hpoId;
		String frequencyTermId;
	}

	public static LinkedHashMap<String, String> loadOrdoFrequencyAnnotations(String orphanetProduct4Path)
			throws IOException, ParserConfigurationException, SAXException {
		LinkedHashMap<String, String> frequencies = new LinkedHashMap<String, String>();

		DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
		DocumentBuilder documentBuilder = factory.newDocumentBuilder();
		Document document = documentBuilder.parse(orphanetProduct4Path);
		Element root = document.getDocumentElement();
		NodeList disorderLists = root.getElementsByTagName("HPODisorderSetStatusList");

		for (int i = 0; i < disorderLists.getLength(); i++) {
			Element disorderList = (Element) disorderLists.item(i);
			NodeList disorders = disorderList.getElementsByTagName("Disorder");
			for (int j = 0; j < disorders.getLength(); j++) {
				Element disorder = (Element) disorders.item(j);
				Element orphaCode = (Element) disorder.getElementsByTagName("OrphaCode").item(0);
				if (orphaCode == null) {
					continue;
				}

				NodeList hpoIds = disorder.getElementsByTagName("HPOId");
				NodeList hpoFrequencies = disorder.getElementsByTagName("HPOFrequency");
				for (int k = 0; k < Math.min(hpoIds.getLength(), hpoFrequencies.getLength()); k++) {
					Element hpoIdElement = (Element) hpoIds.item(k);
					Element hpoFrequencyElement = (Element) hpoFrequencies.item(k);
					String frequencyLabel = extractFrequencyLabel(hpoFrequencyElement);
					if (frequencyLabel == null) {
						continue;
					}

					String key = orphaCode.getTextContent().trim() + "\t" + normalizeHpoId(hpoIdElement.getTextContent());
					if (!frequencies.containsKey(key)) {
						frequencies.put(key, frequencyLabel);
					}
				}
			}
		}

		return frequencies;
	}

	public static LinkedHashMap<String, String> loadManualPhenotypeAssociations(String phenotypeHpoaPath, String diseasePrefix) throws IOException {
		LinkedHashMap<String, String> manualAssociations = new LinkedHashMap<String, String>();

		try (BufferedReader reader = RdfBuildSupport.openUtf8Reader(phenotypeHpoaPath)) {
			String line;
			while ((line = reader.readLine()) != null) {
				if (line.startsWith("#")) {
					continue;
				}

				String[] split = line.split("\t");
				if (split.length <= 3 || !split[0].startsWith(diseasePrefix + ":")) {
					continue;
				}

				String diseaseId = split[0].substring((diseasePrefix + ":").length()).trim();
				String hpoId = normalizeHpoId(split[3]);
				String key = diseaseId + "\t" + hpoId;
				if (!manualAssociations.containsKey(key)) {
					manualAssociations.put(key, "Manual");
				}
			}
		}

		return manualAssociations;
	}

	private static String extractFrequencyLabel(Element hpoFrequencyElement) {
		if (hpoFrequencyElement == null) {
			return null;
		}
		NodeList names = hpoFrequencyElement.getElementsByTagName("Name");
		if (names.getLength() == 0) {
			return null;
		}
		return names.item(0).getTextContent().trim();
	}

	private static String normalizeHpoId(String value) {
		return value.trim().replace("HP:", "");
	}

	private static LinkedHashMap<String, String> createOrdoFrequencyMap() {
		LinkedHashMap<String, String> map = new LinkedHashMap<String, String>();
		map.put("Obligate (100%)", "0040280");
		map.put("Very frequent (99-80%)", "0040281");
		map.put("Frequent (79-30%)", "0040282");
		map.put("Occasional (29-5%)", "0040283");
		map.put("Very rare (<4-1%)", "0040284");
		map.put("Excluded (0%)", "0040285");
		return map;
	}

	/** annotation source blank node에 들어갈 작성자와 페이지 정보를 생성한다. */
	public static AnnotationSource createAnnotationSource(String creator, String page) {
		AnnotationSource source = new AnnotationSource();
		source.creator = creator;
		source.page = page;
		return source;
	}

	public static void writeOrdoPhenotypeAssociationTtl(
			String outputPath,
			Map<String, String> manualAssociations,
			Map<String, String> frequencyByAssociation,
			AnnotationSource source) throws IOException {
		try (BufferedWriter writer = RdfBuildSupport.createBufferedWriter(outputPath)) {
			writer.write("PREFIX dcterms: <http://purl.org/dc/terms/>"); writer.newLine();
			writer.write("PREFIX foaf: <http://xmlns.com/foaf/0.1>"); writer.newLine();
			writer.write("PREFIX hoom: <http://www.semanticweb.org/ontology/HOOM#>"); writer.newLine();
			writer.write("PREFIX oa: <http://www.w3.org/ns/oa#>"); writer.newLine();
			writer.write("PREFIX obo: <http://purl.obolibrary.org/obo/>"); writer.newLine();
			writer.write("PREFIX ordo: <http://www.orpha.net/ORDO/>"); writer.newLine();
			writer.write("PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>"); writer.newLine();
			writer.write("PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>"); writer.newLine();

			int blankNodeCounter = 0;
			for (OrdoPhenotypeAnnotation annotation : buildOrdoAnnotations(manualAssociations, frequencyByAssociation)) {
				writer.write("<https://pubcasefinder.dbcls.jp/phenotype_context/disease:ORDO:" + annotation.ordoId + "/phenotype:HP:" + annotation.hpoId + ">");
				writer.newLine();
				writer.write("    a oa:Annotation ;");
				writer.newLine();
				writer.write("    oa:hasTarget ordo:Orphanet_" + annotation.ordoId + " ;");
				writer.newLine();
				writer.write("    oa:hasBody obo:HP_" + annotation.hpoId + " ;");
				writer.newLine();
				if (annotation.frequencyTermId != null) {
					writer.write("    hoom:with_frequency obo:HP_" + annotation.frequencyTermId + " ;");
					writer.newLine();
				}
				writer.write("    dcterms:source _:b" + (++blankNodeCounter) + " ;");
				writer.newLine();
				writer.write("    obo:ECO_9000001 obo:ECO_0000218 .");
				writer.newLine();

				writer.write("_:b" + blankNodeCounter);
				writer.newLine();
				writer.write("    dcterms:creator \"" + source.creator + "\" ;");
				writer.newLine();
				writer.write("    foaf:page <" + source.page + "> .");
				writer.newLine();
			}

			writeFrequencyLabels(writer);
		}
	}

	public static void writeManualPhenotypeAssociationTtl(
			String outputPath,
			String diseaseNamespaceInPath,
			String diseaseResourcePrefix,
			String diseasePrefixLine,
			Map<String, String> manualAssociations,
			AnnotationSource source) throws IOException {
		try (BufferedWriter writer = RdfBuildSupport.createBufferedWriter(outputPath)) {
			writer.write("PREFIX dcterms: <http://purl.org/dc/terms/>"); writer.newLine();
			writer.write("PREFIX foaf: <http://xmlns.com/foaf/0.1>"); writer.newLine();
			writer.write(diseasePrefixLine); writer.newLine();
			writer.write("PREFIX oa: <http://www.w3.org/ns/oa#>"); writer.newLine();
			writer.write("PREFIX obo: <http://purl.obolibrary.org/obo/>"); writer.newLine();
			writer.write("PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>"); writer.newLine();
			writer.write("PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>"); writer.newLine();

			int blankNodeCounter = 0;
			for (String key : manualAssociations.keySet()) {
				String[] split = key.split("\t");
				String diseaseId = split[0];
				String hpoId = split[1];

				writer.write("<https://pubcasefinder.dbcls.jp/phenotype_context/disease:" + diseaseNamespaceInPath + ":" + diseaseId + "/phenotype:HP:" + hpoId + ">");
				writer.newLine();
				writer.write("    a oa:Annotation ;");
				writer.newLine();
				writer.write("    oa:hasTarget " + diseaseResourcePrefix + diseaseId + " ;");
				writer.newLine();
				writer.write("    oa:hasBody obo:HP_" + hpoId + " ;");
				writer.newLine();
				writer.write("    dcterms:source _:b" + (++blankNodeCounter) + " ;");
				writer.newLine();
				writer.write("    obo:ECO_9000001 obo:ECO_0000218 .");
				writer.newLine();

				writer.write("_:b" + blankNodeCounter);
				writer.newLine();
				writer.write("    dcterms:creator \"" + source.creator + "\" ;");
				writer.newLine();
				writer.write("    foaf:page <" + source.page + "> .");
				writer.newLine();
			}
		}
	}

	private static Iterable<OrdoPhenotypeAnnotation> buildOrdoAnnotations(
			Map<String, String> manualAssociations,
			Map<String, String> frequencyByAssociation) {
		LinkedHashMap<String, OrdoPhenotypeAnnotation> annotations = new LinkedHashMap<String, OrdoPhenotypeAnnotation>();
		for (String key : manualAssociations.keySet()) {
			String[] split = key.split("\t");
			OrdoPhenotypeAnnotation annotation = new OrdoPhenotypeAnnotation();
			annotation.ordoId = split[0];
			annotation.hpoId = split[1];
			String frequencyLabel = frequencyByAssociation.get(key);
			annotation.frequencyTermId = ORDO_FREQUENCY_TO_HPO.get(frequencyLabel);
			annotations.put(key, annotation);
		}
		return annotations.values();
	}

	private static void writeFrequencyLabels(BufferedWriter writer) throws IOException {
		for (Map.Entry<String, String> entry : ORDO_FREQUENCY_TO_HPO.entrySet()) {
			writer.write("obo:HP_" + entry.getValue());
			writer.newLine();
			writer.write("    rdfs:label \"" + entry.getKey() + "\"@en .");
			writer.newLine();
		}
	}
}
